webapp/publisher/snaps/views.py

```diff
@@ -1345,42 +1351,90 @@ def get_snap_builds(snap_name):
     context["github_user"] = github.get_user()
 
     if context["github_user"]:
         context["github_repositories"] = github.get_user_repositories()
 
-    context["snap_builds_enabled"] = len(lp_snaps) >= 1
-    if context["snap_builds_enabled"]:
-        context["snap_builds"] = []
+    # Get built snap in launchpad with this store name
+    lp_snap = launchpad.get_snap_by_store_name(details["snap_name"])
+
+    if lp_snap:
+        # Git repository without GitHub hostname
+        context["github_repository"] = lp_snap["git_repository_url"][19:]
+
         bsi_url = flask.current_app.config["BSI_URL"]
 
-        for build in lp_snaps[0].builds:
-            link = build_link(bsi_url, lp_snaps[0], build)
+        builds = launchpad.get_collection_entries(
+            lp_snap["builds_collection_link"]
+        )
+
+        context["snap_builds_enabled"] = bool(builds)
+
+        for build in builds:
+            link = build_link(bsi_url, lp_snap, build)
             status = map_build_and_upload_states(
-                build.buildstate, build.store_upload_status
+                build["buildstate"], build["store_upload_status"]
             )
 
             context["snap_builds"].append(
                 {
-                    "id": build.self_link.split("/")[-1],
-                    "arch_tag": build.arch_tag,
-                    "datebuilt": build.datebuilt,
-                    "duration": build.duration,
+                    "id": build["self_link"].split("/")[-1],
+                    "arch_tag": build["arch_tag"],
+                    "datebuilt": build["datebuilt"],
+                    "duration": build["duration"],
                     "link": link,
-                    "logs": build.build_log_url,
-                    "revision_id": build.revision_id,
+                    "logs": build["build_log_url"],
+                    "revision_id": build["revision_id"],
                     "status": status,
-                    "title": build.title,
+                    "title": build["title"],
                 }
             )
 
     return flask.render_template("publisher/builds.html", **context)
+
+
+@publisher_snaps.route("/<snap_name>/builds", methods=["POST"])
+@login_required
+def get_snap_builds_post(snap_name):
+    try:
+        details = api.get_snap_info(snap_name, flask.session)
+    except ApiResponseErrorList as api_response_error_list:
+        if api_response_error_list.status_code == 404:
+            return flask.abort(404, "No snap named {}".format(snap_name))
+        else:
+            return _handle_error_list(api_response_error_list.errors)
+    except ApiError as api_error:
+        return _handle_errors(api_error)
+
+    # Get built snap in launchpad with this store name
+    lp_snap = launchpad.get_snap_by_store_name(details["snap_name"])
+
+    if lp_snap:
+        flask.flash(
+            "It is not possible to change the git repository for this snap.",
```

Comment:
Huh? Really? Why not?

Reply:
Basically this form should be done just once by the user, this is a protection to check if the user somehow is trying to do the POST request twice.

The reason behind is because we are creating a Snap in launchpad. I agreed with @Lukewh to just let the user do it once. If we want to allow the user to change the repo multiple times then we will have to delete the snap from launchpad and create a new one (or to update the snap, I don't know), but we are not sure about the implications of doing that but it could be something that we want to allow in the future.

Reply:
Hmm. If we were sure the user submitted the form twice, we should say that in the message, or just silently ignore the second submission.

If we're not sure that's what happened, it seems like we should expose the error a bit more clearly.

How sure are we? Or can we tell the difference?

Reply:
Oh wait...

If we got the response from the API, couldn't we check if the repo URL against that snap is actually the same as the one being submitted? If it was, then it seems safe to just continue happily. This should cover the double-submission problem.

If the repo being submitted is different, then we could raise a real error, rather than the `flask.flash` version. Or we could simply go ahead and call `launchpad.new_snap` and let the API respond with an error, which we can hopefully leave to error naturally (see my other comment about `raise_for_status`)

Reply:
The API will create a new snap because the md5 hash of the git repo is different.

Reply:
It will raise an error only when using the same repo

Reply:
If there is a Snap in launchpad with this store_name if because it has already a git repo so we shouldn't create another one because in launchpad the store_name is not unique. What we could do is only show the error when the git repo is different and do nothing if the git repo is the same one. WDYT?

Reply:
> It will raise an error only when using the same repo

This is kinda why I was confused by the `name` being a hash of the repo. The way I would expect this to work would be that you would tell it 2 things - the snap name and the repository you want to build from.

What happens if you make the `name` attribute the same as the `snap_name` instead of being generated from the `repo`?

Reply:
It will create a Snap in launchpad with this name. If we want to change this behaviour what happen with all the snaps already created in launchpad using the hash of another property? It will create an inconsistency in launchpad, for this feature everything will work as expected but I don't know what could go wrong if we decide to change that property.

Reply:
@jkfran please file an issue about tidying up our usage of the API